### DISCUSSION TITLE
How to build functionalized PEG

### DISCUSSION THREAD

**Discussion by amin-sagar** [marrink-lab/polyply_1.0#296] [Q&A] [answered]
Dear polyply developers,
Thanks for sharing this awesome code.
I am tying to find a way to build molecules like this
![image](https://user-images.githubusercontent.com/118898409/213713996-265bc39a-be4c-4032-b421-5c05cd304fc7.png)
I am thinking of the following protocol.
Make 3 arms separately.
```
polyply gen_seq -o A.json -name PEOA -seq A -from_string A:10:1:PEO-1.0
polyply gen_seq -o B.json -name PEOB -seq B -from_string B:10:1:PEO-1.0
polyply gen_seq -o C.json -name PEOC -seq C -from_string C:10:1:PEO-1.0
```
Generate parameters for the three arms
```
polyply gen_params -lib martini3 -name PEOA -seqf A.json -o A.itp
polyply gen_params -lib martini3 -name PEOB -seqf B.json -o B.itp
polyply gen_params -lib martini3 -name PEOC -seqf C.json -o C.itp
```
Link the three arms - Last residue of first arm - Last residue of second arm, Last residue of first arm - First residue of third arm.
```
polyply gen_seq -f A.itp B.itp C.itp -from_file a:PEOA b:PEOB c:PEOC -seq a b c -connects 0:1:9-9 0:2:9-0 -o link-test1.json -name link-test1
```
Generate parameters for the linked molecule
```
polyply gen_params -name linked -lib martini3 -seqf link-test1.json -o linked.itp
```
Generate coordinates for the linked molecule
```
polyply gen_coords -p system.top -name linked -o linked.gro -box 10 10 10
```
Does this seem like the correct approach?

If yes, Is there a way to modify the last residue of the first arm to N so that it the final molecule there is a Nitrogen atom making three bonds with the three arms? 

I would be really grateful for any suggestions.

Amin.

**Comment by fgrunewald** [comment]
@amin-sagar, 

Hi Amin, 

You're on the right track. You don't even have to generate the itp files for the arms separately.  However, we don't have parameters for the PEG-NH2, PEG-CCOH, and triple Amine PEG interactions in the library at the moment. That means it will not work out of the box. So you have two options: 
1) You could derive them yourself following the some of the steps mentioned [here](https://github.com/marrink-lab/polyply_1.0/issues/190). The bead assignments for the missing groups, we can already share from earlier work. Note that you have to decide on the protonation state of all missing residues as well. 
2) You wait for us to patch the library with these parameters, which will probably happen in the next few weeks, as we have a massive update coming up. 

Let me know what you prefer and we can help either way. 

Cheers,
Fabian

**Comment by amin-sagar** [comment]
Thanks @fgrunewald
I would like to start with option 1, although by the time I figure out things, maybe your massive update will already be there.
However, I anticipate running simulations with other derivatives in the future so this would be really useful for me.

The first step seems to be to run all-atom simulations. I can derive the GAFF parameters using Antechamber to do this.
Would it be better to run the simulations of the whole polymer or just PEG-NH2, PEG-COOH and triply linked PEG?

Best,
Amin.

**Comment by fgrunewald** [comment]
Hey @amin-sagar, 

Option 1 it is. If you run further derivatives in future it for sure makes sense, plus we can extend the polyply library with parameters that are not part of the update. I do my best to guide you through this. 

To answer your question, if you can afford it I'd run the whole molecule, since it is quite small. The parameters extracted from that molecule should be transferable to other derivatives as well. For example, the NH2-PEG-PEG angle will most likely not depend on the length of full molecule. However, you need to mind your protonation states. If you work in water (pH7), I'd say the amine needs to be protonated and the acid deprotonated. The central amine is a bit more tricky. I'd probably run one version where the amines and acid are charged, and one version where they are all neutral. That way you at least have all parameters. 

Once, the simulations are completed we can discuss how to extract the parameters and the mapping. 

Cheers,
Fabian

**Comment by amin-sagar** [comment]
Thanks again @fgrunewald 
I have run 5 independent simulations of 500 ns each for PEG3 solvated in TIP3P water. I ran the simulations with the amines protonated and carboxylic group deprotonated as we are interested in properties at neutral pH. After deriving the parameters for neutral pH, I can try to repeat the procedure with other protonation states, if required.
I might have made a mistake, though. 
Since the molecule is so small, I was able to parameterize the whole molecule with Antechamber and run simulations considering it as a single unit. For the subsequent steps, should I break down the molecule into separate residues, i.e. NH2-PEG, PEG, COOH-PEG and N-triple linker? 
Also, can you please share the bead assignments?
I am really grateful for your help.

Best,
Amin.

**Comment by fgrunewald** [comment]
@amin-sagar 

Sounds good to me! 500ns should be good enough for mapping. Below the mapping:

![mapping](https://user-images.githubusercontent.com/32294573/214892510-bf57b61b-bb1d-4dcf-941d-2b333e51206f.png)


In terms of bead assignment there are some reasonable options. I've annotated them all in addition to the neutral beads (i.e. charged beads have a Q in the name). Ideally you select the best options  based on some property benchmark. Simplest case you select those beads that give you the best Rg value in relation to the 500ns AA data. You might want to define an improper dihedral angle for the central linker (i.e. N-PEG-PEG-PEG).  

In terms of mapping, I would group the same fragments together. For example, the angle NH2-PEG-PEG should be the same in both arms etc. The PEG bond length and angle I would keep as they are in polyply. Most likely they will be different from what you get, but then again Amber, OPLS, CHARMM, and GROMOS all give slightly different parameters for PEG. 

If you have gotten the parameters, I'm also more than happy to add them to the polyply library and help you writing the .ff files. 

Cheers,
Fabian 

P.S For mapping trajectories there are lot's of options. I use [this program](https://github.com/fgrunewald/fast_forward)

**Comment by amin-sagar** [comment]
Thanks @fgrunewald 
I wrote a mapping file like this
```
; Generate CG parameters for PEG3
[ molecule ]
peg PEG3 

[ martini ]
TQ4p SN3r Q2 SP2

[ atoms ]
     1    N01  TQ4p
     2    H01  TQ4p
     3    H02  TQ4p
     4    C02  TQ4p
     5    H03  TQ4p
     6    H04  TQ4p
     7    C03  SN3r
     8    H05  SN3r
     9    H06  SN3r
    10    C04  SN3r
    11    H07  SN3r
    12    H08  SN3r
    13    C05  SN3r
    14    H09  SN3r
    15    H10  SN3r
    16    C06  SN3r
    17    H11  SN3r
    18    H12  SN3r
    19    C07  SN3r
    20    H13  SN3r
    21    H14  SN3r
    22    C08  SN3r
    23    H15  SN3r
    24    H16  SN3r
    25    C09  Q2
    26    H17  Q2
    27    H18  Q2
    28    C10  Q2
    29    H19  Q2
    30    H20  Q2
    31    C11  SN3r
    32    H21  SN3r
    33    H22  SN3r
    34    C12  SN3r
    35    H23  SN3r
    36    H24  SN3r
    37    C13  SN3r
    38    H25  SN3r
    39    H26  SN3r
    40    C14  SN3r
    41    H27  SN3r
    42    H28  SN3r
    43    C15  SN3r
    44    H29  SN3r
    45    H30  SN3r
    46    C16  SN3r
    47    H31  SN3r
    48    H32  SN3r
    49    C17  TQ4p
    50    H33  TQ4p
    51    H34  TQ4p
    52    N18  TQ4p
    53    H35  TQ4p
    54    H36  TQ4p
    55    C19  Q2
    56    H37  Q2
    57    H38  Q2
    58    C20  SN3r
    59    H39  SN3r
    60    H40  SN3r
    61    C21  SN3r
    62    H41  SN3r
    63    H42  SN3r
    64    C22  SN3r
    65    H43  SN3r
    66    H44  SN3r
    67    C23  SN3r
    68    H45  SN3r
    69    H46  SN3r
    70    C24  SN3r
    71    H47  SN3r
    72    H48  SN3r
    73    C25  SN3r
    74    H49  SN3r
    75    H50  SN3r
    76    C26  SN3r
    77    H51  SN3r
    78    H52  SN3r
    79    O27  SN3r
    80    O28  SN3r
    81    O29  SN3r
    82    N30  Q2
    83    O31  SN3r
    84    O32  SN3r
    85    O33  SN3r
    86    O34  SN3r
    87    O35  SN3r
    88    O36  SN3r
    89    C37  SP2
    90    O38  SP2
    91    O39  SP2
    92    H53  SP2
```
and ran 
ff_map -f PEG3_prod1_nopbc.xtc -s PEG3_prod1.tpr -mf PEG3.map -o PEG3_CG.xtc -mols peg
This gave me a mapped CG trajectory.
I am not sure about the next step. 
What's the format of the itp file?
The example on Fast-Forward is 
```
[ bonds ]
1 2 1 2000 ; group1
2 3 1 2000 ; group1
4 5 1 1000
```
I think the first two numbers are the atom numbers having the bond. What do the third and fourth numbers represent?

And, to be sure, in the POPC example POPC was treated as a single residue which makes sense for POPC. However, in my case, I should get 4 residues - NPEG, CPEG, PEG and NLIN (Central linker). Should I define these in my input files or can this be done after  all the steps are done? Currently my all atom simulations define the entire molecule as a single residue.

Thanks again for your help.
Amin.

**Comment by fgrunewald** [comment]
Hi Amin, 

You're on a good track but let me clarify some things about the mapping.

- in the mapping file that is used for fast forward you entered the bead-types. However, you need to enter the bead names at the Martini level. Within a single residue each bead needs to have a unique name. 
- This brings me to your other question, if you should split  the AA molecule into multiple residues or keep as one. It is easier to keep it as one at the AA level for now but you need to regenerate the mapping file. 
- You can use [this website](https://github.com/jbarnoud/cgbuilder) to generate the mapping file with the correct and unique atom names that is required for fast forward. Note that you will have to remove the `[to]` directive and add the residue name directive.
- Before you generate the mapped trajectory make sure to pbc complete it using 'gmx trjconv -f trajectory.xtc -o pbc.xtc -pbc whole'. 
- Next step: generate a mapped .gro file and a mapped .xtc and view them using a trajectory viewer (e.g. VMD) to make sure that it makes sense. For example, there should't be any free floating beads etc. 

With respect to the itp-file, I would recommend to skim [this part](https://manual.gromacs.org/current/reference-manual/topologies/topology-file-formats.html) and [this bit](https://manual.gromacs.org/current/reference-manual/functions/bonded-interactions.html) of the gromacs manual first. 

Once you have done this you need to generate an template itp-file for your molecule. You can already do this with polyply. You just have to add 3 blocks and their links with PEG. Here is a [tutorial](https://github.com/marrink-lab/polyply_1.0/wiki/Tutorial:-writing-.ff-input-files) which shows the syntax to write the .ff files. Next generate the residue graph. Treat the terminal residues and the central residue, simply as a single block like you have done with the PEGs above (e.g. `-from_string A:1:1:NH2ter-1.0`). Your command should then looks something like below. Note the residue names are NH2ter for the terminal NH2 group, NT for the central linker bead, and ACDter for the terminal acid. The sequence is the two PEG arms first and then the central bead and then the last PEG arm and ACD. 

```
polyply gen_seq -from_string PEG10:10:1:PEO-1. NHT:1:1:NH2ter-1 NT:1:1:NT-1. ACD:1:1:ACDter-1. -seq NHT PEG10 NHT PEG10 NT PEG10 ACD -connects 0:1:0-0 2:3:0-0 1:4:9-0 3:4:9-0 4:5:0-0 5:6:9-0 -o link-test1.json -name link-test1
```
You can draw the molecule using the following small code:
```
import sys
import networkx as nx
import matplotlib.pyplot as plt
from polyply.src.simple_seq_parsers import parse_json

graph = parse_json(sys.argv[1])
nx.draw_networkx(graph, pos=nx.kamada_kawai_layout(graph))
plt.show()
```
Next generate the itp file. 
```
polyply gen_params -name linked -lib martini3 -f YOUR_LINKS.ff -seqf link-test1.json -o linked.itp
```

**Comment by amin-sagar** [comment]
Thanks @fgrunewald 
I think I understand how to write the map file better now.
I generated it using CGBuilder
It looks like this now
```
[ molecule ]
peg PEG3

[ martini ]
B0 B1 B2 B3 B4 B5 B6 B7 B8 B9 B10 B11 B12 

[ atoms ]
1	N01	B0
2	H01	B0
3	H02	B0
4	C02	B0
5	H03	B0
6	H04	B0
7	C03	B1
8	H05	B1
9	H06	B1
10	C04	B1
11	H07	B1
12	H08	B1
13	C05	B2
14	H09	B2
15	H10	B2
16	C06	B2
17	H11	B2
18	H12	B2
19	C07	B3
20	H13	B3
21	H14	B3
22	C08	B3
23	H15	B3
24	H16	B3
25	C09	B4
26	H17	B4
27	H18	B4
28	C10	B4
29	H19	B4
30	H20	B4
31	C11	B9
32	H21	B9
33	H22	B9
34	C12	B9
35	H23	B9
36	H24	B9
37	C13	B10
38	H25	B10
39	H26	B10
40	C14	B10
41	H27	B10
42	H28	B10
43	C15	B11
44	H29	B11
45	H30	B11
46	C16	B11
47	H31	B11
48	H32	B11
49	C17	B12
50	H33	B12
51	H34	B12
52	N18	B12
53	H35	B12
54	H36	B12
55	C19	B4
56	H37	B4
57	H38	B4
58	C20	B5
59	H39	B5
60	H40	B5
61	C21	B5
62	H41	B5
63	H42	B5
64	C22	B6
65	H43	B6
66	H44	B6
67	C23	B6
68	H45	B6
69	H46	B6
70	C24	B7
71	H47	B7
72	H48	B7
73	C25	B7
74	H49	B7
75	H50	B7
76	C26	B8
77	H51	B8
78	H52	B8
79	O27	B1
80	O28	B2
81	O29	B3
82	N30	B4
83	O31	B9
84	O32	B10
85	O33	B11
86	O34	B5
87	O35	B6
88	O36	B7
89	C37	B8
90	O38	B8
91	O39	B8
92	H53	B8
```
I generated the .json file which looks like this.

![residue_graph](https://user-images.githubusercontent.com/118898409/216114857-1a6f6aa3-53d9-4487-a371-8192dc9e3ce3.png)

I am trying to write the .ff file now. I have tried to write the atoms and bonds section.
Does this seem to be going in the right direction?
```
[ moleculetype ]
;molname  nrexcl
PEG 1
[ atoms ]
   ;id  type resnr residue atom cgnr charge
	1   TQ4p  1     PEG     B0   1   +1
	2   SN3r  1     PEG     B1   2   0
	3   SN3r  1     PEG     B2   3   0
	4   SN3r  1     PEG     B3   4   0
	5   SN3a  1     PEG     B4   5   0
	6   SN3r  1     PEG     B5   6   0
	7   SN3r  1     PEG     B6   7   0
	8   SN3r  1     PEG     B7   8   0
	9   SP2   1     PEG     B8   9   -1
	10  SN3r  1     PEG     B9   10  0
	11  SN3r  1     PEG     B10  11  0
	12  SN3r  1     PEG     B11  12  0
	13  TQ4p  1     PEG     B12  13  +1
[ bonds ]
1 2 1
2 3 1
3 4 1
4 5 1
5 6 1
6 7 1
7 8 1
8 9 1
4 9 1
9 10 1
10 11 1
11 12 1
```

**Reply by fgrunewald** [reply]
Overall sounds good. Is the last file the itp file you generate or the ff-file you wrote?

**Comment by amin-sagar** [comment]
@fgrunewald 
Sorry for the delay in my response.
The file I posted was actually my attempt of writing a .ff file for input.
I wrote the following .itp file by hand and gave it to Swarm-CG for optimization
```
[ moleculetype ]
;molname  nrexcl
PEG3 1
[ atoms ]
   ;id  type resnr residue atom cgnr charge
	1   TN6d  1     PEG     B0   1   0
	2   SN3r  1     PEG     B1   2   0
	3   SN3r  1     PEG     B2   3   0
	4   SN3r  1     PEG     B3   4   0
	5   SN3a  1     PEG     B4   5   0
	6   SN3r  1     PEG     B5   6   0
	7   SN3r  1     PEG     B6   7   0
	8   SN3r  1     PEG     B7   8   0
	9   SP2   1     PEG     B8   9   0
	10  SN3r  1     PEG     B9   10  0
	11  SN3r  1     PEG     B10  11  0
	12  SN3r  1     PEG     B11  12  0
	13  TN6d  1     PEG     B12  13  0
[ bonds ]
1  2 1 0 0 ; group 1
12 13  1 0 0 ; group 1

2  3  1 0 0 ; group 2
3  4  1 0 0 ; group 2
6  7  1 0 0 ; group 2
7  8  1 0 0 ; group 2
10 11 1 0 0 ; group 2
11 12 1 0 0 ; group 2

4  5  1 0 0 ; group 3
5  6  1 0 0 ; group 3
5  10 1 0 0 ; group 3

8  9 1 0 0 ; group 4

[ angles ]
1 2 3 2 0 0 ; A1
11 12 13 2 0 0 ; A1

2 3 4 2 0 0 ; A2
6 7 8 2 0 0 ; A2
10 11 12 2 0 0 ; A2

3 4 5 2 0 0 ; A3
5 6 7 2 0 0 ; A3
5 10 11 2 0 0 ; A3

4 5 10 2 0 0 ; A4
4 5 6 2 0 0 ; A4
6 5 10 2 0 0 ; A4 

7 8 9 2 0 0 ; A5
```
The optimized results look this this.
![optimized_CG_model_distributions](https://user-images.githubusercontent.com/118898409/218131927-de3c9b9a-eb42-4374-adba-cfd71670bb5a.png)
It doesn't seem to like bimodal distributions.
The output itp is as follows
```
[ moleculetype ]
; molname        nrexcl
PEG3             1


[ atoms ]
; id type resnr residue   atom  cgnr    charge     mass

1    TN6d     1     PEG     B0  1      0.00000     36.00
2    SN3r     1     PEG     B1  2      0.00000     54.00
3    SN3r     1     PEG     B2  3      0.00000     54.00
4    SN3r     1     PEG     B3  4      0.00000     54.00
5    SN3a     1     PEG     B4  5      0.00000     54.00
6    SN3r     1     PEG     B5  6      0.00000     54.00
7    SN3r     1     PEG     B6  7      0.00000     54.00
8    SN3r     1     PEG     B7  8      0.00000     54.00
9     SP2     1     PEG     B8  9      0.00000     54.00
10   SN3r     1     PEG     B9  10     0.00000     54.00
11   SN3r     1     PEG    B10  11     0.00000     54.00
12   SN3r     1     PEG    B11  12     0.00000     54.00
13   TN6d     1     PEG    B12  13     0.00000     36.00


[ bonds ]
;   i     j   funct   length   force.c.

; bond type 1
    1     2       1    0.279  11803.43           ; 1
   12    13       1    0.279  11803.43           ; 1

; bond type 2
    2     3       1    0.328  4700.04           ; 2
    3     4       1    0.328  4700.04           ; 2
    6     7       1    0.328  4700.04           ; 2
    7     8       1    0.328  4700.04           ; 2
   10    11       1    0.328  4700.04           ; 2
   11    12       1    0.328  4700.04           ; 2

; bond type 3
    4     5       1    0.346  6757.86           ; 3
    5     6       1    0.346  6757.86           ; 3
    5    10       1    0.346  6757.86           ; 3

; bond type 4
    8     9       1    0.345  6485.54           ; 4


[ angles ]
;   i     j     k   funct     angle   force.c.

; angle type 1
    1     2     3       2    129.75     44.16           ; 1
   11    12    13       2    129.75     44.16           ; 1

; angle type 2
    2     3     4       2    131.23     22.27           ; 2
    6     7     8       2    131.23     22.27           ; 2
   10    11    12       2    131.23     22.27           ; 2

; angle type 3
    3     4     5       2    134.33     22.35           ; 3
    5     6     7       2    134.33     22.35           ; 3
    5    10    11       2    134.33     22.35           ; 3

; angle type 4
    4     5    10       2    113.04     12.39           ; 4
    4     5     6       2    113.04     12.39           ; 4
    6     5    10       2    113.04     12.39           ; 4

; angle type 5
    7     8     9       2    145.54     13.16           ; 5
```
These are the results for the simulations where all the ends were neutral and in this case, I used the only available option for terminal beads.
I have run ff_inter as well and the extracted distributions from the mapped trajectory look similar although the plots from Swarm-CG look a bit smoothened, maybe just for plotting.
I have three questions/requests.
1) Do you have any suggestions about do to retrieve the bimodal distributions from the coarse-grained model.
2) Should I include dihedral angles for optimization. Your previous results seem to suggest that it's not straightforward as it can influence the stability of simulations (https://pubs.acs.org/doi/10.1021/acs.jpcb.8b04760) 
3) Can you please give me an example of how to write [link] segments for this molecule? For example, how do I define how to link the central linker with PEO?
I am really grateful for your help.

**Reply by fgrunewald** [reply]
Hi Armin, 

No worries. I'm happy to help even though I'm currently at BPS in San Diego, so my response time will be a bit slower than usual. 

1. In general I would say the matches look not too bad. There is really nothing you can do about bimodal distributions at the moment. 
2. As long as you keep the PEG dihedrals from within polyply I think you only need to define 1 improper dihedral `5 4 6 10`. That should be fairly stable. We'll see. 
3. The link and molecule files for this molecule are very simple. For example, if you use the ff file below with the residue graph it gives you all bonds between the center residue and the 3 attached PEG residues. In your case each residue is a single bead. 
```
[ moleculetype ]
; name nexcl.
PEO         1
;
[ atoms ]
1  SN3r    1   PEO   EC  1   0.000  45
[ moleculetype ]
; name nexcl.
NT         1
;
[ atoms ]
1  SN3a    1   NT   NT  1   0.000  45
[ link ]
resname "NT|PEO"
[ bonds ]
NT >EC 1 0.346  6757.86  
[ link ]
resname "NT|PEO"
[ bonds ]
EC >NT 1 0.346  6757.86  
```

**Reply by fgrunewald** [reply]
Note that the '>' refers to the relative resid. For example, the first link applies when PEO has a resid higher than NT and the second when PEO has a resid lower than NT

**Comment by amin-sagar** [comment]
@fgrunewald 
Thanks once again.
I ran the optimization including the improper dihedral.
![optimized_CG_model_distributions](https://user-images.githubusercontent.com/118898409/221984525-9b5225b2-69c1-430f-a3a3-0f47a69cfa09.png)
The optimized parameters look like this.
```
[ moleculetype ]
; molname        nrexcl
PEG3             1


[ atoms ]
; id type resnr residue   atom  cgnr    charge     mass

1    TN6d     1     PEG     B0  1      0.00000     36.00
2    SN3r     1     PEG     B1  2      0.00000     54.00
3    SN3r     1     PEG     B2  3      0.00000     54.00
4    SN3r     1     PEG     B3  4      0.00000     54.00
5    SN3a     1     PEG     B4  5      0.00000     54.00
6    SN3r     1     PEG     B5  6      0.00000     54.00
7    SN3r     1     PEG     B6  7      0.00000     54.00
8    SN3r     1     PEG     B7  8      0.00000     54.00
9     SP2     1     PEG     B8  9      0.00000     54.00
10   SN3r     1     PEG     B9  10     0.00000     54.00
11   SN3r     1     PEG    B10  11     0.00000     54.00
12   SN3r     1     PEG    B11  12     0.00000     54.00
13   TN6d     1     PEG    B12  13     0.00000     36.00


[ bonds ]
;   i     j   funct   length   force.c.

; bond type 1
    1     2       1    0.279  11943.75           ; 1
   12    13       1    0.279  11943.75           ; 1

; bond type 2
    2     3       1    0.327  6906.09           ; 2
    3     4       1    0.327  6906.09           ; 2
    6     7       1    0.327  6906.09           ; 2
    7     8       1    0.327  6906.09           ; 2
   10    11       1    0.327  6906.09           ; 2
   11    12       1    0.327  6906.09           ; 2

; bond type 3
    4     5       1    0.344  8571.20           ; 3
    5     6       1    0.344  8571.20           ; 3
    5    10       1    0.344  8571.20           ; 3

; bond type 4
    8     9       1    0.347  6302.51           ; 4


[ angles ]
;   i     j     k   funct     angle   force.c.

; angle type 1
    1     2     3       2    131.29     40.03           ; 1
   11    12    13       2    131.29     40.03           ; 1

; angle type 2
    2     3     4       2    131.45     34.06           ; 2
    6     7     8       2    131.45     34.06           ; 2
   10    11    12       2    131.45     34.06           ; 2

; angle type 3
    3     4     5       2    128.66     32.49           ; 3
    5     6     7       2    128.66     32.49           ; 3
    5    10    11       2    128.66     32.49           ; 3

; angle type 4
    4     5    10       2    114.94      8.84           ; 4
    4     5     6       2    114.94      8.84           ; 4
    6     5    10       2    114.94      8.84           ; 4

; angle type 5
    7     8     9       2    133.47     18.49           ; 5


[ dihedrals ]
;   i     j     k     l   funct     dihedral   force.c.   mult.

; dihedral type 1
    5     4     6    10       4       179.26    1.77       0     ; 1
```
I created the following ff file following your example and I can use it to generate the itp file and initial coordinates.
```
[ moleculetype ]
; name nexcl.
PEO         1
;
[ atoms ]
1  SN3r    1   PEO   EC  1   0.000  45
[ moleculetype ]
; name nexcl.
NT         1
;
[ atoms ]
1  SN3a    1   NT   NT  1   0.000  54
[ link ]
resname "NT|PEO"
[ bonds ]
NT >EC 1 0.346  6757.86  
[ dihedrals ]
[ link ]
resname "NT|PEO"
[ bonds ]
EC >NT 1 0.346  6757.86  
[ dihedrals ]
[ moleculetype ]
; name nexcl.
NH2ter     1
;
[ atoms ]
1  TQ4p    1   NH2ter   NH2ter  1   0.000  36
[ link ]
resname "NH2ter|PEO"
[ bonds ]
NH2ter >EC 1 0.279  11803.43
[ link ]
resname "NH2ter|PEO"
[ bonds ]
EC >NH2ter 1 0.279  11803.43

[ moleculetype ]
; name nexcl.
ACDter     1
;
[ atoms ]
1  SP2    1   ACDter   ACDter  1   0.000  54
[ link ]
resname "ACDter|PEO"
[ bonds ]
ACDter >EC 1 0.345  6485.54
[ link ]
resname "ACDter|PEO"
[ bonds ]
EC >ACDter 1 0.345  6485.54
```
I have two issues now.
1) I tried looking at the [ link ] sections of the other molecules but I am not sure how to write the [ dihedrals ] section for NT.
Can you please help me with this?
2) Should I change the masses of the beads according to the atoms that they represent?

Best,
Amin.

**Reply by fgrunewald** [reply]
Hi Amin, 

To answer your questions:

1. The syntax for the dihedral following the residue graph we used above should be:
```
[link]
resname "PEO|NT"
[dihedrals]
NT <EC >EC >>EC <your parameters>
```
Please note that the ordering in the original residue graph we used and the SwarmCG itp are not the same. This is not an issue it just means the itps are not 1 to 1 comparable. 
2. No you should use the Martini masses as defined in the interaction table. In fact I also need to change the PEG mass to 54 instead of 45. It does not influence any simulation results but it is better to be consistent.

**Comment by amin-sagar** [comment]
@fgrunewald 
Everything seems to work including the bonds, angles between the new residues and PEO, except the improper dihedral.
My ff file (part about NT residue) is as follows
```
[ moleculetype ]
; name nexcl.
PEO         1
;
[ atoms ]
1  SN3r    1   PEO   EC  1   0.000  54

[ moleculetype ]
; name nexcl.
NT         1
;
[ atoms ]
1  SN3a    1   NT   NT  1   0.000  54
[ link ]
resname "NT|PEO"
[ bonds ]
NT >EC 1 0.346  6757.86  
[ link ]
resname "NT|PEO"
[ bonds ]
EC >NT 1 0.346  6757.86  
[ link ]
resname "NT|PEO"
[ angles ]
EC >EC >>NT 2    138.36     26.40
[ link ]
resname "NT|PEO"
[ angles ]
NT >EC >>EC 2    138.36     26.40
[ link ]
resname "NT|PEO"
[ dihedrals ]
NT <EC >EC >>EC  2    1.97    2.16 
```
If I generate an itp file using the command.
```polyply gen_params -name linked -lib martini3 -f testff2.ff -seqf link-test1.json -o linked.itp```
The output itp file has a lot of dihedrals where the first atom is NT (atom number 23) but then all the atoms with id < id of NT followed by atoms with id > id of NT followed by atoms with id > id of third atom. Like follows (partial list)

```
23  2 25 27 2 1.97 2.16
23  2 25 28 2 1.97 2.16
23  2 26 28 2 1.97 2.16
23  2 25 29 2 1.97 2.16
23  2 26 29 2 1.97 2.16
23  2 27 29 2 1.97 2.16
23  2 25 30 2 1.97 2.16
23  2 26 30 2 1.97 2.16
23  2 27 30 2 1.97 2.16
23  2 28 30 2 1.97 2.16
23  2 25 31 2 1.97 2.16
23  2 26 31 2 1.97 2.16
23  2 27 31 2 1.97 2.16
23  2 28 31 2 1.97 2.16
23  2 29 31 2 1.97 2.16
23  2 25 32 2 1.97 2.16
23  2 26 32 2 1.97 2.16
23  2 27 32 2 1.97 2.16
23  2 28 32 2 1.97 2.16
23  2 29 32 2 1.97 2.16
23  2 30 32 2 1.97 2.16
23  2 25 33 2 1.97 2.16
23  2 26 33 2 1.97 2.16
23  2 27 33 2 1.97 2.16
23  2 28 33 2 1.97 2.16
23  2 29 33 2 1.97 2.16
23  2 30 33 2 1.97 2.16
23  2 31 33 2 1.97 2.16
23  3 25 27 2 1.97 2.16
23  3 25 28 2 1.97 2.16
23  3 26 28 2 1.97 2.16
23  3 25 29 2 1.97 2.16
23  3 26 29 2 1.97 2.16
23  3 27 29 2 1.97 2.16
23  3 25 30 2 1.97 2.16
```
It's be easy to manually edit the itp file but it would be nice to know what I am doing wrong.
Best,
Amin.

**Reply by fgrunewald** [reply]
@amin-sagar could you try the following link instead:
```
[ link ]
resname "NT|PEO"
[ impropers ]
NT <EC >EC >>EC  2    1.97    2.16 
[ edges ]
NT <EC
NT >EC
NT >>EC
```

**Reply by amin-sagar** [reply]
@fgrunewald 
This gives no dihedrals (except for the ones for PEG) in the generated itp.

**Comment by fgrunewald** [comment]
Hi, you print the output of running polyply with the '-vv' option. Also
what are the resids of the 4 atoms involved in the dihedral?

amin-sagar ***@***.***> schrieb am Do., 9. März 2023, 16:53:

**Reply by amin-sagar** [reply]
Hi. I get a very long list saying
DEBUG - step - polyply.src.apply_links - Found no matchs for node.
I have attached the output.
[gen_params.log](https://github.com/marrink-lab/polyply_1.0/files/10933640/gen_params.log)

**Reply by amin-sagar** [reply]
Sorry, I forgot to write that the 4 residues are 23,11,22,24 (NT,PEO,PEO,PEO).

**Comment by fgrunewald** [comment, marked as answer]
That's the issue then. I messed up the relative ordering. The signa in
front indicate the relative order of resids.

In this case it should be 'NT <<EC <EC >EC'. Update but keep the edges.


amin-sagar ***@***.***> schrieb am Do., 9. März 2023, 17:16:

**Reply by amin-sagar** [reply]
@fgrunewald Thanks a lot.
This seems to work.
The final segment is
```
[ link ]
resname "NT|PEO"
[ dihedrals ]
NT <<EC <EC >EC  2    1.97    2.16
[ edges ]
NT <EC
NT >EC
NT <<EC
```

**Comment by fgrunewald** [comment]
@amin-sagar If everything works nicely feel free to open a PR to contribute some of the parameters to the polyply library for others to use.
